Evaluate using HomeAssistant rules in ruff

> If it makes things simpler then please make a PR that changes from bandit to flake8-bandit.
> 
> Can you please ( if you have not already) the rules from HA, commented out, that way it is easier to activate when you find t8mr.


_Originally posted by @janiversen in https://github.com/pymodbus-dev/pymodbus/issues/1476#issuecomment-1500557392_

https://github.com/home-assistant/core/blob/7eccef87c2cccb834f5ff394f765e697831390e3/pyproject.toml#L240


- [ ] "B007", # Loop control variable {name} not used within loop body
- [ ] "B014", # Exception handler with duplicate exception
- [x] "C",  # complexity
- [x] "D",  # docstrings
- [x] "E",  # pycodestyle
- [x] "F",  # pyflakes/autoflake
- ~~"ICN001", # import concentions; {name} should be imported as {asname}~~
- [x] "PGH004",  # Use specific rule codes when using noqa
- [x] "PLC0414", # Useless import alias. Import alias does not rename original package.
- [x] "SIM105", # Use contextlib.suppress({exception}) instead of try-except-pass
- [x] "SIM117", # Merge with-statements that use the same scope
- [x] "SIM118", # Use {key} in {dict} instead of {key} in {dict}.keys()
- [x] "SIM201", # Use {left} != {right} instead of not {left} == {right}
- [x] "SIM212", # Use {a} if {a} else {b} instead of {b} if not {a} else {a}
- [x] "SIM300", # Yoda conditions. Use 'age == 42' instead of '42 == age'.
- [x] "SIM401", # Use get from dict with default instead of an if block
- ~~"T20",  # flake8-print~~
- [ ] "TRY004", # Prefer TypeError exception for invalid type
- [x] "RUF006", # Store a reference to the return value of asyncio.create_task
- [x] "UP",  # pyupgrade
- [x] "W",  # pycodestyle

ignore = [

- [x] "D202",  # No blank lines allowed after function docstring
- ~~"D203",  # 1 blank line required before class docstring~~
- ~~"D213",  # Multi-line docstring summary should start at the second line~~
- ~~"D406",  # Section name should end with a newline~~
- ~~"D407",  # Section name underlining~~
- [x] "E501",  # line too long
- [x] "E731",  # do not assign a lambda expression, use a def
Ignored due to performance: https://github.com/charliermarsh/ruff/issues/2923
- ~~"UP038", # Use `X | Y` in `isinstance` call instead of `(X, Y)`~~

> > we already use bandit directly and mccabe via pylint, so it puzzles me why those 2 are big jobs....the third one is new to me.
> 
> For McCabe/pylint, I intend to actually reduce the complexity and not just use `pylint: ignore` :)

If https://github.com/pymodbus-dev/pymodbus/pull/1510 is OK, we're down to two functions with a McCabe complexity of > 10.

> pymodbus/transaction.py:127:9: C901 `execute` is too complex (23 > 10)
> pymodbus/transaction.py:335:9: C901 `_recv` is too complex (19 > 10)

Note that both of them have large `if/else if/else` blocks that are functionally switch statements.  If 
https://github.com/charliermarsh/ruff/issues/3597 gets implemented that will reduce their scores.

--- a/ruff.toml
+++ b/ruff.toml
@@ -14,6 +14,7 @@ ignore = [
 ]
 line-length = 120
 select = [
+    "B007",   # Loop control variable {name} not used within loop body
     "D",      # docstrings
     "E",      # pycodestyle errors
     "W",      # pycodestyle warnings

--- a/ruff.toml
+++ b/ruff.toml
@@ -15,6 +15,8 @@ ignore = [
 line-length = 120
 select = [
     "B007",   # Loop control variable {name} not used within loop body
+    "B014",   # Exception handler with duplicate exception
+    "C",      # complexity
     "D",      # docstrings
     "E",      # pycodestyle errors
     "W",      # pycodestyle warnings
@@ -32,8 +34,8 @@ select = [
     "PLC",    # pylint
     "PGH",    # pygrep-hooks
     # "TRY",    # tryceratops
+    "TRY004", # Prefer TypeError exception for invalid type
     # "B",      # bandit
-    "C",      # complexity
 ]
 [pydocstyle]
 convention = "pep257"

--- a/test/test_factory.py
+++ b/test/test_factory.py
@@ -121,7 +121,7 @@ class TestFactory:
 
     def test_response_working(self):
         """Test a working response factory decoders"""
-        for func, msg in self.response:
+        for _func, msg in self.response:
             self.client.decode(msg)
 
     def test_response_errors(self):
@@ -134,7 +134,7 @@ class TestFactory:
 
     def test_requests_working(self):
         """Test a working request factory decoders"""
-        for func, msg in self.request:
+        for _func, msg in self.request:
             self.server.decode(msg)
 
     def test_client_factory_fails(self):

